File: vms/c/ic.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#ifdef _WIN32
#   include <fcntl.h>
#   include <io.h>
#endif // _WIN32

int *mem = NULL;
int mem_size = 0;

int ip = 0;
int rb = 0;

void resize_mem(int addr) {
    if (addr >= mem_size) {
        int old_mem_size = mem_size;
        while (addr >= mem_size) mem_size <<= 1;
        mem = (int *)realloc(mem, mem_size * sizeof(int));
        memset(mem + old_mem_size, 0, (mem_size - old_mem_size) * sizeof(int));
    }
}

int get_mem(int addr) {
    resize_mem(addr);
    return mem[addr];
}

void set_mem(int addr, int val) {
    resize_mem(addr);
    mem[addr] = val;
}

const int MODE_MUL[] = { 100, 1000, 10000 };

int get_param(int idx) {
    int mode = get_mem(ip) / MODE_MUL[idx] % 10;
    switch (mode) {
        case 0: // position mode
            return get_mem(get_mem(ip + idx + 1));
        case 1: // immediate mode
            return get_mem(ip + idx + 1);
        case 2: // relative mode
            return get_mem(rb + get_mem(ip + idx + 1));
        default:
            fprintf(stderr, "mode error: ip %d idx %d\n", ip, idx);
            exit(1);
    }
}

void set_param(int idx, int val) {
    int mode = get_mem(ip) / MODE_MUL[idx] % 10;
    switch (mode) {
        case 0: // position mode
            set_mem(get_mem(ip + idx + 1), val);
            break;
        case 2: // relative mode
            set_mem(rb + get_mem(ip + idx + 1), val);
            break;
        default:
            fprintf(stderr, "mode error: ip %d idx %d\n", ip, idx);
            exit(1);
    }
}

int get_input() {
    return getc(stdin);
}

void set_output(int val) {
    putc(val, stdout);
    fflush(stdout);
}

void run() {
    while (true) {
        int oc = get_mem(ip) % 100;

        switch (oc) {
            case 1: // add
                set_param(2, get_param(0) + get_param(1));
                ip += 4;
                break;
            case 2: // mul
                set_param(2, get_param(0) * get_param(1));
                ip += 4;
                break;
            case 3: { // in
                int value = get_input();
                if (value == EOF) {
                    fprintf(stderr, "no more inputs\n");
                    exit(1);
                }
                set_param(0, value);
                ip += 2;
                break;
            }
            case 4: { // out
                int value = get_param(0);
                ip += 2;
                set_output(value);
                break;
            }
            case 5: // jnz
                if (get_param(0) != 0) {
                    ip = get_param(1);
                } else {
                    ip += 3;
                }
                break;
            case 6: // jz
                if (get_param(0) == 0) {
                    ip = get_param(1);
                } else {
                    ip += 3;
                }
                break;
            case 7: // lt
                set_param(2, get_param(0) < get_param(1) ? 1 : 0);
                ip += 4;
                break;
            case 8: // eq
                set_param(2, get_param(0) == get_param(1) ? 1 : 0);
                ip += 4;
                break;
            case 9: // arb
                rb += get_param(0);
                ip += 2;
                break;
            case 99: // hlt
                return;
            default:
                fprintf(stderr, "opcode error: ip %d oc %d\n", ip, oc);
                exit(1);
        }
    }
}

int main(int argc, char **argv) {
#ifdef _WIN32
    _setmode(_fileno(stdin), _O_BINARY);
    _setmode(_fileno(stdout), _O_BINARY);
    _setmode(_fileno(stderr), _O_BINARY);
#endif // _WIN32

    mem_size = 64;
    mem = (int*)malloc(mem_size * sizeof(int));
    memset(mem, 0, mem_size * sizeof(int));

    FILE *input = fopen(argv[1], "rt");

    int idx = 0;
    char ch = ',';
    int num = 0;

    while (ch == ',') {
        if (fscanf(input, "%d%c", &num, &ch) != 2) {
            fprintf(stderr, "parse error");
            exit(1);
        }
        set_mem(idx++, num);
    }

    run();

    return 0;
}

File: vms/rust/ic.rs
use std::env;
use std::fs;
use std::io;
use std::io::prelude::*;
use std::process::exit;

struct Vm {
    ip: usize,
    rb: i32,
    mem: Vec<i32>
}

impl Vm {
    pub fn new() -> Vm {
        Vm {
            ip: 0,
            rb: 0,
            mem: vec![0; 64]
        }
    }

    fn get_mem(self: &Vm, addr: usize) -> i32 {
        if addr >= self.mem.len() {
            0
        } else {
            self.mem[addr]
        }
    }

    fn set_mem(self: &mut Vm, addr: usize, val: i32) {
        if addr >= self.mem.len() {
            let mut new_size = self.mem.len();
            while addr >= new_size {
                new_size <<= 1;
            }
            self.mem.resize(new_size, 0);
        }
        self.mem[addr] = val
    }

    const MODE_MUL: [i32; 3] = [100, 1000, 10000];

    fn get_param(self: &Vm, idx: usize) -> Result<i32, String> {
        match self.get_mem(self.ip) / Vm::MODE_MUL[idx] % 10 {
            0 => Ok(self.get_mem(self.get_mem(self.ip + idx + 1).try_into().unwrap())), // position mode
            1 => Ok(self.get_mem(self.ip + idx + 1)), // immediate mode
            2 => Ok(self.get_mem((self.rb + self.get_mem(self.ip + idx + 1)).try_into().unwrap())), // relative mode
            _ => Err(format!("mode error: ip {} idx {}", self.ip, idx))
        }
    }

    fn set_param(self: &mut Vm, idx: usize, val: i32) -> Result<(), String> {
        match self.get_mem(self.ip) / Vm::MODE_MUL[idx] % 10 {
            0 => Ok(self.set_mem(self.get_mem(self.ip + idx + 1).try_into().unwrap(), val)), // position mode
            2 => Ok(self.set_mem((self.rb + self.get_mem(self.ip + idx + 1)).try_into().unwrap(), val)), // relative mode
            _ => Err(format!("mode error: ip {} idx {}", self.ip, idx))
        }
    }

    pub fn run(self: &mut Vm, get_input: fn() -> i32, set_output: fn(i32)) -> Result<(), String> {
        loop {
            let oc = self.get_mem(self.ip) % 100;

            match oc {
                1 => { // add
                    self.set_param(2, self.get_param(0)? + self.get_param(1)?)?;
                    self.ip += 4;
                },
                2 => { // mul
                    self.set_param(2, self.get_param(0)? * self.get_param(1)?)?;
                    self.ip += 4;
                },
                3 => { // in
                    self.set_param(0, get_input())?;
                    self.ip += 2;
                },
                4 => { // out
                    set_output(self.get_param(0)?);
                    self.ip += 2;
                },
                5 => { // jnz
                    match self.get_param(0)? {
                        0 => self.ip += 3,
                        _ => self.ip = self.get_param(1)?.try_into().unwrap()
                    };
                },
                6 => { // jz
                    match self.get_param(0)? {
                        0 => self.ip = self.get_param(1)?.try_into().unwrap(),
                        _ => self.ip += 3
                    };
                },
                7 => { // lt
                    self.set_param(2, if self.get_param(0)? < self.get_param(1)? { 1 } else { 0 })?;
                    self.ip += 4;
                },
                8 => { // eq
                    self.set_param(2, if self.get_param(0)? == self.get_param(1)? { 1 } else { 0 })?;
                    self.ip += 4;
                },
                9 => { // arb
                    self.rb += self.get_param(0)?;
                    self.ip += 2;
                },
                99 => { // hlt
                    return Ok(());
                }
                _ => {
                    return Err(format!("opcode error: ip {} oc {}", self.ip, oc));
                }
            }
        }
    }
}

#[cfg(windows)]
const ENDL: &'static str = "\r\n";
#[cfg(not(windows))]
const ENDL: &'static str = "\n";

fn get_input() -> i32 {
    let mut buf: [u8; 1] = [0];
    match io::stdin().read_exact(&mut buf) {
        Ok(()) => (),
        Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => {
            eprint!("no more inputs{}", ENDL);
            exit(1);
        },
        Err(e) => {
            panic!("{}", e);
        }
    }
    buf[0] as i32
}

fn set_output(val: i32) {
    let val = u8::try_from(val).expect("output value out of range");
    print!("{}", val as char);
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let path = args.get(1)
        .expect("no program specified");

    let data = fs::read_to_string(path)
        .expect("error reading program file");
    let data = data.trim_end().split(',');

    let mut vm = Vm::new();

    for (idx, item) in data.enumerate() {
        let item: i32 = item.parse().expect("invalid program");
        vm.set_mem(idx, item);
    }

    vm.run(get_input, set_output).unwrap();
}
